### PR TITLE
Adding the PreflightCheck interface and adding the version check preflight check

### PR DESCRIPTION
**Pull request by ytimocin**
# Description

This pull request introduces a preflight check system for validating upgrade operations in the Radius project. The changes include the implementation of a `Registry` to manage checks, a `VersionCompatibilityCheck` to validate version upgrade paths, and associated tests to ensure reliability. Additionally, the documentation has been updated to explain the new system.

### Preflight Check System Implementation:

* **`pkg/upgrade/preflight/registry.go`**: Introduced a `Registry` to manage and execute preflight checks, including methods for adding checks (`AddCheck`) and running them (`RunChecks`). The system logs results and halts on critical errors.
* **`pkg/upgrade/preflight/types.go`**: Defined the `PreflightCheck` interface and supporting types like `CheckSeverity` and `CheckResult` to standardize checks and their results.

### Version Compatibility Check:

* **`pkg/upgrade/preflight/version_check.go`**: Added `VersionCompatibilityCheck` to enforce upgrade rules, such as preventing downgrades and skipping multiple versions. Includes a utility function `ValidateVersionJump` for external use.

### Testing:

* **`pkg/upgrade/preflight/registry_test.go`**: Added unit tests for the `Registry` class, covering scenarios like empty registries, successful checks, and error handling for critical failures.
* **`pkg/upgrade/preflight/version_check_test.go`**: Added tests for `VersionCompatibilityCheck` to validate upgrade paths and ensure proper handling of edge cases like downgrades and unresolved versions.

### Documentation:

* **`pkg/upgrade/README.md`**: Updated to include details about the preflight check system, its extensibility, and usage examples for developers.

## Type of change
- This pull request adds or changes features of Radius and has an approved issue (issue link required).
Fixes: #issue_number

## Contributor checklist
Please verify that the PR meets the following requirements, where applicable:

- An overview of proposed schema changes is included in a linked GitHub issue.
    - [ ] Yes
    - [ ] Not applicable
- A design document PR is created in the [design-notes repository](https://github.com/radius-project/design-notes/), if new APIs are being introduced.
    - [ ] Yes
    - [ ] Not applicable
- The design document has been reviewed and approved by Radius maintainers/approvers.
    - [ ] Yes
    - [ ] Not applicable
- A PR for the [samples repository](https://github.com/radius-project/samples) is created, if existing samples are affected by the changes in this PR.
    - [ ] Yes
    - [ ] Not applicable
- A PR for the [documentation repository](https://github.com/radius-project/docs) is created, if the changes in this PR affect the documentation or any user facing updates are made.
    - [ ] Yes
    - [ ] Not applicable
- A PR for the [recipes repository](https://github.com/radius-project/recipes) is created, if existing recipes are affected by the changes in this PR.
    - [ ] Yes
    - [ ] Not applicable

### CODE REVIEW

File: pkg/upgrade/preflight/version_check_test.go [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+			name:           "valid upgrade to next minor version",
+			currentVersion: "v0.43.0",
+			targetVersion:  "v0.44.0",
+			expectSuccess:  true,
+			expectMessage:  "Upgrade from v0.43.0 to v0.44.0 is valid",
+		},
+		{
+			name:           "valid upgrade to latest",
+			currentVersion: "v0.43.0",
+			targetVersion:  "latest",
```
Comment (ytimocin) [inline pkg/upgrade/preflight/version_check_test.go:45]: An incremental check will be added to this part later.

**Review by willdavsmith — APPROVED** [review]
